sys/core/src/Purpose.hs

```diff
@@ -163,45 +165,62 @@ goalAfter allsyncs outsyncs TxsDDefs.ActQui (gid,btree)  =  do
        runStateT (Behave.behAfterAct allsyncs btree qacts) envb1
      writeEnvBtoEnvC envb2
      case (maybt1,maybt2) of
-      (Nothing ,Nothing ) -> return (gid,[])
-      (Just bt1,Nothing ) -> return (gid,bt1)
-      (Nothing ,Just bt2) -> return (gid,bt2)
-      (Just bt1,Just bt2) -> return (gid,bt1++bt2)
+      (Nothing ,Nothing ) -> return (gid,Left [])
+      (Just bt1,Nothing ) -> return (gid,Left bt1)
+      (Nothing ,Just bt2) -> return (gid,Left bt2)
+      (Just bt1,Just bt2) -> return (gid,Left (bt1++bt2))
+goalAfter _ _ _ goal@(_, Right _) = return goal
+      
 
 -- ----------------------------------------------------------------------------------------- --
 -- purpVerdict :  output of  hit/miss verdicts
 
-purpVerdict :: IOC.IOC ()
+purpVerdict :: IOC.IOC Bool -- did any goal hit?
 purpVerdict = do
   envc <- get
   case IOC.state envc of
     IOC.Testing { IOC.purpsts = purpsts } -> do
-      IOC.putMsgs [EnvData.TXS_CORE_USER_INFO $ "Goals: " ++ show (length purpsts)]
-      mapM_ goalVerdict purpsts
+      hitsAndNewPurps <- mapM goalVerdict purpsts
+      let (hasHit, newPurps) = unzip3to2 hitsAndNewPurps
+      IOC.modifyCS (\s -> s {IOC.purpsts = newPurps})
+      return $ or hasHit
     _ -> do
       IOC.putMsgs [EnvData.TXS_CORE_SYSTEM_ERROR "purpVerdict incorrectly used"]
-      return ()
-
-goalVerdict :: (TxsDefs.GoalId,BTree.BTree) -> IOC.IOC ()
-goalVerdict (gid,btree)  =  do
+      return False
+  where
+    unzip3to2 :: [(a,b,c)] -> ([a],[(b,c)])
```

Comment:
Can't we just avoid having this recursion boilerplate by defining:

```haskell
unzip3to2 = foldr (\(x, y, z) (xs, yzs) -> (x:xs, (y, z):yzs)) ([], [])
```

Reply:
yes, that version of unzip3to2 looks nicer.